### ISSUE TITLE
Cannot just gather all interfaces in a package

### ISSUE THREAD

**Issue by moldowan** [lukehutch/fast-classpath-scanner#30]
Imagine you have a package with only interfaces.

Now try to find all these interfaces with FastClasspathScanner:
- .getNamesOfAllClasses() after just .scan() without any Processors returns only classes, not interfaces
- .matchSubclassesOf(Object.class, ...) + .scan() does not match anything
- .matchSubinterfacesOf(Object.class, ...) / .matchClassesImplementing() + .scan() both yield an exception

.matchClassesWithAnnotation() does match the interfaces but my interfaces don't have an annotation.

Expectations:
As there is no .getNamesOfAllInterfaces(), I expect .getNamesOfAllClasses() to return the interface names (besides "regular" class names).
In the end I need the class objects of all interfaces, not the just the names. So for me the best would be if matchSubclassesOf(Object.class, ...) also matched interfaces. This would be consistent with Object.class.isAssignableFrom(MyInterface.class).

**Comment by lukehutch** [comment]
All good points, thanks for the report. I'll take a look at implementing your suggested changes soon.

**Comment by moldowan** [comment]
Thanks! Great tool by the way.

**Comment by lukehutch** [comment]
Thanks! I get so little feedback that I never know until bugs are filed if anyone's even using it! I see 121 stars, which is encouraging, but maybe I should interpret the lack of chatter as a signal that it's generally working for people? I'd be curious to start gathering some data though on how it's being used -- what's your general use case?

**Comment by lukehutch** [comment]
Responding to your points:
* Good catch, getNamesOfAllClasses() is supposed to return not just standard classes, but also interfaces and annotations. Fixed this (will check in shortly). I also added getNamesOfAllStandardClasses(), getNamesOfAllInterfaceClasses() and getNamesOfAllAnnotationClasses().
* I forgot that I actually specifically removed Object.class as a superclass of all classes (since it didn't seem useful), and forgot to update the documentation. (That is why .matchSubclassesOf(Object.class, ...) + .scan() does not match anything.) Fixed the docs, since there are other ways to accomplish this now, as described in the previous bullet point.
* I'll check on the exceptions in your third bullet point, but can you please provide more details in another bug report on what the exception is, and how exactly to trigger it? I'll push another release out once I have fixed this.

Thanks for this bug report!

**Comment by lukehutch** [comment]
Ah, actually I think I understand your third bullet point. Object.class is the superclass of all interfaces, but it's not the superinterface of anything, since it's not an interface. So .matchSubinterfacesOf(Object.class, ...) should throw an exception that Object is not an interface. Is this the exception you're referring to? Also, .matchClassesImplementing(Object.class, ...) + .scan() has the same issue (I assume this is what you meant, since you only mentioned ".matchClassesImplementing() + .scan()"?)

If so, no need to open another bug report, unless you think this is not expected behavior.

**Comment by lukehutch** [comment]
Please test the updated API and let me know if this works for you now. I addition to the fixes and additions to the getAll...() methods described above, see also the new ClassEnumerationMatchProcessor and matchAll...() methods referenced in the commit above.

**Comment by moldowan** [comment]
I will try to test the updated API today but perhaps I have to postpone this until early january.

Regarding the third bullet point:
I did not expect either approach to work, this was just trial and error and I'm fine with the observed behaviour.

Regarding my general use case:
We use FastClasspathScanner in a medium/large sized Java EE project for "compliance testing", like ensuring the existence of certain annotations on a set of classes etc.
Those tests are something that is otherwise very hard or even impossible to cover with tools like Checkstyle etc.

**Comment by moldowan** [comment]
I cloned your repository and executed mvn clean package which failed with:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project fast-classpath-scanner: Compilation failure: Compilation failure:
[ERROR] /c:/Develop/_dev/fast-classpath-scanner/src/main/java/io/github/lukehutch/fastclasspathscanner/matchprocessor/ClassEnumerationMatchProcessor.java:[32,2] cannot find symbol
[ERROR] symbol: class FunctionalInterface
[ERROR] /c:/Develop/_dev/fast-classpath-scanner/src/main/java/io/github/lukehutch/fastclasspathscanner/matchprocessor/SubclassMatchProcessor.java:[32,2] cannot find symbol
[ERROR] symbol: class FunctionalInterface
[ERROR] /c:/Develop/_dev/fast-classpath-scanner/src/main/java/io/github/lukehutch/fastclasspathscanner/matchprocessor/StaticFinalFieldMatchProcessor.java:[72,2] cannot find symbol
[ERROR] symbol: class FunctionalInterface
[ERROR] /c:/Develop/_dev/fast-classpath-scanner/src/main/java/io/github/lukehutch/fastclasspathscanner/matchprocessor/FileMatchProcessor.java:[35,2] cannot find symbol
[ERROR] symbol: class FunctionalInterface
[ERROR] /c:/Develop/_dev/fast-classpath-scanner/src/main/java/io/github/lukehutch/fastclasspathscanner/matchprocessor/ClassAnnotationMatchProcessor.java:[32,2] cannot find symbol
[ERROR] symbol: class FunctionalInterface
[ERROR] /c:/Develop/_dev/fast-classpath-scanner/src/main/java/io/github/lukehutch/fastclasspathscanner/matchprocessor/SubinterfaceMatchProcessor.java:[32,2] cannot find symbol
[ERROR] symbol: class FunctionalInterface
[ERROR] /c:/Develop/_dev/fast-classpath-scanner/src/main/java/io/github/lukehutch/fastclasspathscanner/matchprocessor/InterfaceMatchProcessor.java:[32,2] cannot find symbol
[ERROR] symbol: class FunctionalInterface
[ERROR] /c:/Develop/_dev/fast-classpath-scanner/src/main/java/io/github/lukehutch/fastclasspathscanner/matchprocessor/FileMatchContentsProcessor.java:[34,2] cannot find symbol
[ERROR] symbol: class FunctionalInterface
```
Environment:
```
c:\Develop\_dev\fast-classpath-scanner>mvn -version
Apache Maven 3.2.5 (12a6b3acb947671f09b81f49094c53f426d8cea1; 2014-12-14T18:29:23+01:00)
Maven home: C:\Program Files\apache-maven-3.2.5
Java version: 1.7.0_80, vendor: Oracle Corporation
Java home: C:\Develop\CCS\jdk1.7.0_80\jre
Default locale: de_DE, platform encoding: Cp1252
OS name: "windows 7", version: "6.1", arch: "amd64", family: "windows"
```

**Comment by lukehutch** [comment]
Can you please test this works on with Java 8?I'm not sure how I broke Java
7 compatibility, but I'm not seeing this...

**Comment by jforge** [comment]
Functional Interfaces are a Java 8 features, you can't compile this using Java 7 unless you at least "emulate" this missing Annotation.
Or you get rid of Java 7 compatibility (maybe except for a specific branch for legacy Java 7 support).

To use Functional Interfaces is such a good idea, that I would prefer not to switch back ;-)

**Comment by lukehutch** [comment]
Right, I know they're a Java 8 thing, but for some reason that I actually
never understood, this project was compiling and building fine for me with
just Java 7 (and for others, including the guy who first submitted JDK7
compatability patches). I figured that they must have backported the
@FunctionalInterface annotation to JDK7 or something. I didn't actually
change anything to do with that annotation in the recent checkins.

**Comment by lukehutch** [comment]
I looked into this, and there's no way I can see to make the project build under both Java 7 and Java 8 without some Maven trickery that adds alternative versions of the classes to the classpath depending on the JDK version. However, the project builds fine under JDK 1.8 in Java 7 compatibility mode, which is how I build the project for Maven Central (i.e. the jar built this way does include the FunctionalInterface annotations, but these are ignored if the jar is used with a Java 7 project). I updated the README.md to reflect these Java 7 compatibility issues, thanks for pointing them out.

PS you shouldn't need to build this yourself, you can just use a pre-built jar, assuming I fixed the API properly.

**Comment by moldowan** [comment]
.matchAllInterfaceClasses() is working as expected in 1.9.11. Thank you!

**Comment by lukehutch** [comment]
Great! Thanks for testing the changes and letting me know.

**Comment by lukehutch** [comment]
PS You might also find the new classgraph visualization feature I added in
1.9.11 useful.

**Comment by lukehutch** [comment]
Hi, I just released 1.9.12, but I wanted to let you know I renamed ClassEnumerationMatchProcessor to ClassMatchProcessor, since I needed to use it for generic class handling for tasks other than just enumeration, to avoid creating even more MatchProcessors in future. Sorry to change this on you.
